### PR TITLE
Nomad Constructor Jobs and Dispatch

### PR DESCRIPTION
**Pull request by dadgar**
This PR introduces:

* Nomad Constructor Jobs which act as batch job templates
* `nomad dispatch` command and API to create an instance of a constructor job.

### CODE REVIEW

**Comment by jippi** [comment]
@dadgar what kind of job is `Constructor Jobs` - similar  to`docker run` adhoc jobs or?

File: command/job_dispatch.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+	// Truncate the id unless full length is requested
+	length := shortId
+	if verbose {
+		length = fullId
+	}
+
+	// Check that we got exactly one node
+	args = flags.Args()
+	if l := len(args); l < 1 || l > 2 {
```
Comment (dadgar) [inline command/job_dispatch.go]: It can be one or two but no less and no more :)

File: client/task_runner_test.go [hunk 2]
```diff
@@ -1244,3 +1246,66 @@ func TestTaskRunner_VaultManager_Signal(t *testing.T) {
...
+		"exit_code": "0",
+		"run_for":   "1s",
+	}
+	fileName := "test"
+	task.DispatchInput = &structs.DispatchInputConfig{
+		File: fileName,
+	}
+	alloc.Job.Constructor = &structs.ConstructorConfig{}
+
+	// Add an encrypted payload
```
Comment (dadgar) [inline client/task_runner_test.go]: Yeah good catch!

File: command/job_dispatch.go (hunk 1 above)
Comment (jippi) [inline command/job_dispatch.go]: the comment says "exactly one node" ? :)

**Pull request merged**